### PR TITLE
BUG: Raise if histogram cannot create finite bin sizes

### LINKED ISSUE #27142: BUG: zero-width histogram bins if the data values are in a small range close to numeric precision
### Describe the issue:

`np.histogram` can produce zero-width bins if the data values are in a small range close to numeric precision. Noted in https://github.com/matplotlib/matplotlib/issues/28685

I'm aware that there is no reasonable representation / that calculating a reasonable representation close to the numeric precision is not possible. But would it be an option to check the bin-width and error out instead of returning nonsense histogram binning?

### Reproduce the code example:

```python
In [1]: import numpy as np

In [2]: a = np.array([1, 1+2e-16] * 10)

In [3]: counts, lims = np.histogram(a, bins=10)

In [4]: counts
Out[4]: array([ 0, 10,  0,  0,  0,  0,  0,  0,  0, 10])

In [5]: lims
Out[5]: array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [6]: np.diff(lims)
Out[6]: 
array([0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 2.22044605e-16, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00])
```

### Error message:

_No response_

### Python and NumPy Versions:

numpy 1.26.4
python 3.12.2

### Runtime Environment:

_No response_

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by timhoffm**
When many bins are requested in a small value region, it may not be possible to create enough distinct bin edges due to limited numeric precision. Up to now, `histogram` then returned identical subsequent bin edges, which would mean a bin width of 0. These bins could also have counts associated with them.

Instead of returning such unlogical bin distributions, this PR raises a value error if the calculated bins do not all have a finite size.

Closes #27142.

### CODE REVIEW

File: numpy/lib/tests/test_histograms.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def test_object_array_of_0d(self):
             histogram, [np.array(0.4) for i in range(10)] + [np.inf])
 
         # these should not crash
-        np.histogram([np.array(0.5) for i in range(10)] + [.500000000000001])
+        np.histogram([np.array(0.5) for i in range(10)] + [.500000000000002])
```
Comment (timhoffm) [inline numpy/lib/tests/test_histograms.py:273]: Note: This test was failing because one of the created bins had zero width. I assume this was not the intention of the test. It was added in https://github.com/numpy/numpy/pull/10268, which is about coercing values of object arrays. I've increased the value minimally to not run into the zero-width bin case.

**Comment by ngoldbaum** [comment]
Not sure if the pypy failure is real.

**Comment by timhoffm** [comment]
> Not sure if the pypy failure is real.

Can't tell. But even if it's real, it seems unrelated to the PR, because it's an assertion rewrite problem in `test_simd.py`.

**Review by mattip — APPROVED** [review]
Makes sense to me, and tests (including the new one) are passing. This even caught an edge case in an existing test.

**Comment by ngoldbaum** [comment]
The pypy failure went away when I re-triggered it so it must be flake.

Thanks @timhoffm!

**Comment by djhoese** [comment]
Sorry for commenting on a merged PR, but I'm not sure this is issue worthy. I'm having trouble consistently producing the error message added here. If I copy the test case then I get the error. If I increase the base number of the arrays then it works fine:

```python
In [31]: np.histogram_bin_edges(np.array([1.0, 1.0 + 2e-16] * 10))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[31], line 1
----> 1 np.histogram_bin_edges(np.array([1.0, 1.0 + 2e-16] * 10))
...
ValueError: Too many bins for data range. Cannot create 10 finite-sized bins.

In [32]: np.histogram_bin_edges(np.array([2.0, 2.0 + 2e-16] * 10))
Out[32]: array([1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5])

```

Why does starting the array with 1.0 give different bins that starting with 2.0?

For context, I'm trying to reproduce one of my unit tests that has started failing with this new error, but copying and pasting the `repr` of the array (`np.array([999.9999989758, 999.9999989758, 999.9999989758, 999.9999989758])`) involved does not trigger the error interactively like it does during the test.
